jmclient/jmclient/storage.py

```diff
@@ -0,0 +1,333 @@
+        self._write_file(magic + enc_data)
+        self._update_data_hash()
+
+    def _load_file(self, password):
+        data = self._read_file()
+        assert len(self.MAGIC_ENC) == len(self.MAGIC_UNENC) == 8
+        magic = data[:8]
+
+        if magic not in (self.MAGIC_ENC, self.MAGIC_UNENC):
+            raise StorageError("File does not appear to be a joinmarket wallet.")
+
+        data = data[8:]
+
+        if magic == self.MAGIC_ENC:
+            if password is None:
+                raise StorageError("Password required to open wallet.")
+            data = self._decrypt_file(password, data)
+        else:
+            assert magic == self.MAGIC_UNENC
+
+        self.data = self._deserialize(data)
+        self._update_data_hash()
+
+    def _write_file(self, data):
+        assert self.read_only is False
+
+        if not os.path.exists(self.path):
+            # newly created storage
+            with open(self.path, 'wb') as fh:
+                fh.write(data)
+            return
+
+        # using a tmpfile ensures the write is atomic
+        tmpfile = '{}.tmp'.format(self.path)
+
+        with open(tmpfile, 'wb') as fh:
+            shutil.copystat(self.path, tmpfile)
+            fh.write(data)
+
+        #FIXME: behaviour with symlinks might be weird
+        shutil.move(tmpfile, self.path)
+
+    def _read_file(self):
+        # this method mainly exists for easier mocking
+        with open(self.path, 'rb') as fh:
+            return fh.read()
+
+    @staticmethod
+    def _serialize(data):
+        return bencoder.bencode(data)
+
+    @staticmethod
+    def _deserialize(data):
+        return bencoder.bdecode(data)
+
+    def _encrypt_file(self, data):
+        if not self.is_encrypted():
+            return data
+
+        iv = get_random_bytes(16)
+        container = {
+            b'enc': {b'salt': self._hash.salt, b'iv': iv},
+            b'data': self._encrypt(data, iv)
+        }
+        return self._serialize(container)
+
+    def _decrypt_file(self, password, data):
+        assert password is not None
+
+        container = self._deserialize(data)
+        assert b'enc' in container
+        assert b'data' in container
+
+        self._hash = self._hash_password(password, container[b'enc'][b'salt'])
+
+        return self._decrypt(container[b'data'], container[b'enc'][b'iv'])
+
+    def _encrypt(self, data, iv):
+        encrypter = pyaes.Encrypter(
+                pyaes.AESModeOfOperationCBC(self._hash.hash, iv=iv))
+        enc_data = encrypter.feed(self.MAGIC_DETECT_ENC + data)
+        enc_data += encrypter.feed()
+
+        return enc_data
+
+    def _decrypt(self, data, iv):
+        decrypter = pyaes.Decrypter(
+                pyaes.AESModeOfOperationCBC(self._hash.hash, iv=iv))
+        try:
+            dec_data = decrypter.feed(data)
+            dec_data += decrypter.feed()
+        except ValueError:
+            # in most "wrong password" cases the pkcs7 padding will be wrong
```

Comment:
As noted on IRC, `pyaes` is not verifying PKCS7 padding properly, see https://github.com/ricmoo/pyaes/blob/master/pyaes/util.py#L51-L60 for reference, see https://github.com/btcsuite/btcd/pull/1108 for earlier discussion (and links to earlier Joinmarket issue).
It almost certainly doesn't matter here, I'll let you check that @undeath , but we could open a PR for pyaes